src/Microsoft.Health.Fhir.Azure/ContainerRegistry/AzureContainerRegistryAccessTokenProvider.cs

```diff
@@ -0,0 +1,183 @@
+            var request = new HttpRequestMessage(HttpMethod.Post, exchangeUri)
+            {
+                Content = new FormUrlEncodedContent(parameters),
+            };
+
+            string refreshToken;
+            try
+            {
+                var refreshTokenResponse = await SendRequestAsync(request, cancellationToken);
+                refreshTokenResponse.EnsureSuccessStatusCode();
+                var refreshTokenText = await refreshTokenResponse.Content.ReadAsStringAsync();
+                dynamic refreshTokenJson = JsonConvert.DeserializeObject(refreshTokenText);
+                refreshToken = (string)refreshTokenJson.refresh_token;
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Failed to exchange ACR refresh token with aad access token.");
+                throw new AzureContainerRegistryTokenException(Resources.CannotGetAcrAccessToken, ex);
+            }
+
+            if (string.IsNullOrEmpty(refreshToken))
+            {
+                _logger.LogError("ACR refresh token is empty.");
+                throw new AzureContainerRegistryTokenException(Resources.CannotGetAcrAccessToken);
+            }
+
+            _logger.LogInformation("Successfully exchanged ACR refresh token.");
+            return refreshToken;
+        }
+
+        private async Task<string> GetAcrAccessToken(Uri accessTokenUri, string refreshToken, CancellationToken cancellationToken)
+        {
+            var parameters = new List<KeyValuePair<string, string>>();
+            parameters.Add(new KeyValuePair<string, string>("grant_type", "refresh_token"));
+            parameters.Add(new KeyValuePair<string, string>("service", accessTokenUri.Host));
+            parameters.Add(new KeyValuePair<string, string>("refresh_token", refreshToken));
+
+            // Add scope for AcrPull role (granted at registry level).
+            parameters.Add(new KeyValuePair<string, string>("scope", "repository:*:pull"));
+            var request = new HttpRequestMessage(HttpMethod.Post, accessTokenUri)
+            {
+                Content = new FormUrlEncodedContent(parameters),
+            };
+
+            string accessToken;
+            try
+            {
+                var accessTokenResponse = await SendRequestAsync(request, cancellationToken);
+                accessTokenResponse.EnsureSuccessStatusCode();
+                var accessTokenText = await accessTokenResponse.Content.ReadAsStringAsync();
+                dynamic accessTokenJson = JsonConvert.DeserializeObject(accessTokenText);
+                accessToken = accessTokenJson.access_token;
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Failed to get ACR access token with ACR refresh token.");
+                throw new AzureContainerRegistryTokenException(Resources.CannotGetAcrAccessToken, ex);
```

Comment:
>AzureContainerRegistryTokenException [](start = 26, length = 36)

Do we need to catch this exception in filter, it doesn't feel like this resource is should be BadRequest but either 500 or 504?

Reply:
Agree. I have updated to return 500 errors on exception and return 400 when the response is 401 or 404